runtime/hashing/common/src/aws/smithy/kotlin/runtime/hashing/Crc32c.kt

```diff
@@ -0,0 +1,676 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.hashing
+
+import aws.smithy.kotlin.runtime.util.InternalApi
+import kotlin.experimental.and
+import kotlin.experimental.xor
+
+@InternalApi
+public abstract class Crc32cBase : HashFunction {
+    override val blockSizeBytes: Int = 4
+    override val digestSizeBytes: Int = 4
+
+    public abstract fun digestValue(): UInt
+}
+
+/**
+ * CRC32C checksum. Note: [digest] will return the bytes (big endian) of the CRC32C integer value. Access [digestValue]
+ * directly to avoid doing the integer conversion yourself.
+ */
+@InternalApi
+public expect class Crc32c() : Crc32cBase
+
+/**
+ * Compute the CRC32C hash of the current [ByteArray]
+ */
+@InternalApi
+public fun ByteArray.crc32c(): UInt = Crc32c().apply { update(this@crc32c) }.digestValue()
+
+public actual class Crc32c : Crc32cBase() {
+    private val md = CRC32CImpl()
+
+    override fun update(input: ByteArray, offset: Int, length: Int): Unit = md.update(input, offset, length)
+
+    override fun digest(): ByteArray {
```

Comment:
This doesn't change for each `actual`, we can push this to `Crc32cBase` (same thing with `Crc32` actually)

Reply:
Would it be smart to simplify further and combine Crc32cBase and Crc32Base?

Reply:
Yeah I think so, good catch